module-system/module-implementations/sov-ibc-transfer/src/call.rs

```diff
@@ -0,0 +1,131 @@
+use ibc::core::ics24_host::identifier::{ChannelId, PortId};
+use ibc::core::timestamp::Timestamp;
+use ibc::core::ExecutionContext;
+use ibc::Signer;
+use sov_state::WorkingSet;
+
+use crate::context::EscrowExtraData;
+use crate::Transfer;
+
+#[derive(borsh::BorshDeserialize, borsh::BorshSerialize, Debug, PartialEq)]
+pub struct SDKTokenTransfer<C: sov_modules_api::Context> {
+    /// the port on which the packet will be sent
+    pub port_id_on_a: PortId,
+    /// the channel by which the packet will be sent
+    pub chan_id_on_a: ChannelId,
+    /// Timeout height relative to the current block height.
+    /// The timeout is disabled when set to None.
+    pub timeout_height_on_b: TimeoutHeight,
+    /// Timeout timestamp relative to the current block timestamp.
+    /// The timeout is disabled when set to 0.
+    pub timeout_timestamp_on_b: Timestamp,
+    /// The address of the token to be sent
+    pub token_address: C::Address,
+    /// The amount of tokens sent
+    pub amount: sov_bank::Amount,
+    /// The address of the token sender
+    pub sender: Signer,
+    /// The address of the token receiver on the counterparty chain
+    pub receiver: Signer,
+    /// Additional note associated with the message
+    pub memo: Memo,
+}
+
+impl<C> Transfer<C>
+where
+    C: sov_modules_api::Context,
+{
+    pub fn transfer(
+        &self,
+        sdk_token_transfer: SDKTokenTransfer<C>,
+        execution_context: &mut impl ExecutionContext,
+        token_ctx: &mut impl TokenTransferExecutionContext<EscrowExtraData<C>>,
+        working_set: Rc<RefCell<&mut WorkingSet<C::Storage>>>,
+    ) -> Result<sov_modules_api::CallResponse> {
+        let msg_transfer: MsgTransfer = {
+            let denom = {
+                let token_name = self
+                    .bank
+                    .get_token_name(
+                        &sdk_token_transfer.token_address,
+                        &mut working_set.borrow_mut(),
+                    )
+                    .ok_or(anyhow::anyhow!(
+                        "Token with address {} doesn't exist",
+                        sdk_token_transfer.token_address
+                    ))?;
+
+                if self.token_was_created_by_ibc(
+                    &token_name,
+                    &sdk_token_transfer.token_address,
+                    &mut working_set.borrow_mut(),
+                ) {
+                    // The token was created by the IBC module, and the ICS-20
+                    // denom was stored in the token name. Hence, we need to use
+                    // the token name as denom.
+                    token_name
+                } else {
+                    // This applies to all other tokens created on this
+                    // sovereign SDK chain. The token name is not guaranteed to
+                    // be unique, and hence we must use the token address (which
+                    // is guaranteed to be unique) as the ICS-20 denom to ensure
```

Comment:
I am seeing, when transferring tokens between two Sovereign chains, we exclusively use the token address as the denom, disregarding the token's name.
Whether the token lives on the source chain, in which case the denom is based on the token address, or if the IBC mints the token on the source chain (which represents a token on the destination chain). In latter, we use a prefixed denom, where the base denom comes from a `MsgRecvPacket`, which was earlier crafted by relayers based on an emitted `TransferEvent` from the destination chain...Where the [denom](https://github.com/cosmos/ibc-rs/blob/c126e7f3ddd75980bb08c2ef7407a4b60189dc77/crates/ibc/src/applications/transfer/relay/send_transfer.rs#L190) is also set as the token address!

Reply:
> when transferring tokens between two Sovereign chains

In the initial version, we will only support sending tokens to Tendermint-based chains, as we will only have a tendermint light client installed.

> In latter, we use a prefixed denom, where the base denom comes from a MsgRecvPacket, which was earlier crafted by relayers based on an emitted TransferEvent from the destination chain...Where the [denom](https://github.com/cosmos/ibc-rs/blob/c126e7f3ddd75980bb08c2ef7407a4b60189dc77/crates/ibc/src/applications/transfer/relay/send_transfer.rs#L190) is also set as the token address!

However, if we *were* to support Sovereign <-> Sovereign transfers, then the denom of tokens we receive would indeed represent the address of the token on the other chain. Note that the token we mint on our side would have a different address, but its denom would be set to the string representation of the address of the token on the other chain. However, for received tokens, we treat the token name as a black box; we never try to interpret it in any way.

Reply:
> the denom of tokens we receive would indeed represent the address of the token on the other chain

Exactly. And looks like it holds true for bunch of other non-Cosmos chains, which makes string reps of token addresses the most common denom.

By the way, the transfer module stores the "denom → token_address" in the `minted_token` field. This allows us to map, e.g. the `channel-1/port-1/0xb79…4f5ea` received from `MsgRecvPacket` to the corresponding token address on the source chain. So, I'm still a bit puzzled about why can't we work with the `MsgTransfer` and include the string reps of the token's address as denom? What required us to define `SDKTokenTransfer` on top of that? 

Anyway, as chatted earlier, Let's dive deeper into these details when writing tests.